cli/doctest_runner.rs

```diff
@@ -0,0 +1,358 @@
+    let path = PathBuf::from(&url.to_string());
+    let media_type = MediaType::from(&path);
+    let specifier = ModuleSpecifier::resolve_url(&url.to_string())?;
+    let parsed_module = ast::parse(&specifier, &source_code, &media_type)?;
+    let mut doc_tester = DocTester::new(parsed_module);
+    results.extend(doc_tester.get_comments());
+  }
+  Ok(results)
+}
+
+pub fn prepare_doctests(
+  jsdocs: Vec<(Location, String)>,
+  fail_fast: bool,
+  quiet: bool,
+  filter: Option<String>,
+) -> Result<String, AnyError> {
+  let mut test_file =  "import * as assert from \"https://deno.land/std@0.70.0/testing/asserts.ts\";\n".to_string();
+  let mut import_set = HashSet::new();
+
+  let cwd = std::env::current_dir()?;
+  let cwd_url_str = Url::from_directory_path(cwd)
+    .map(|url| url.to_string())
+    .unwrap_or_else(|_| "".to_string());
+
+  let tests: String = jsdocs
+    .into_iter()
+    .filter_map(|(loc, example)| {
+      let ex_str = clean_string(&example);
+      let test_tag = extract_test_tag(&ex_str);
+      if test_tag == Some("text") {
+        return  None;
+      }
+      let ignore = test_tag == Some("ignore");
+      extract_imports(&ex_str).into_iter().for_each(|import| { import_set.insert(import.to_string()); });
+      let caption = extract_caption(&ex_str);
+      let code_body = extract_code_body(&ex_str);
+      let is_async = has_await(&ex_str);
+      let res = format!(
+        "\nDeno.test({{\n\tname: \"{} - {} (line {})\",\n\tignore: {},\n\t{} fn() {{\n{}\n}}\n}});\n",
+        loc.filename.replace(&cwd_url_str, ""),
+        caption.unwrap_or(""),
+        loc.line,
+        ignore,
+        if is_async { "async"} else {""},
+        code_body
+    );
+    Some(res)
+    })
+    .collect();
+  let imports_str = import_set.into_iter().collect::<Vec<_>>().join("\n");
+  test_file.push_str(&imports_str);
+  test_file.push_str("\n");
+  test_file.push_str(&tests);
+
+  let options = if let Some(filter) = filter {
+    json!({ "failFast": fail_fast, "reportToConsole": !quiet, "disableLog": quiet, "isDoctest": true, "filter": filter })
+  } else {
+    json!({ "failFast": fail_fast, "reportToConsole": !quiet, "disableLog": quiet, "isDoctest": true })
+  };
+
+  let run_tests_cmd = format!(
+    "\n// @ts-ignore\nDeno[Deno.internal].runTests({});\n",
+    options
+  );
+
+  test_file.push_str(&run_tests_cmd);
+  Ok(test_file)
+}
+
+fn clean_string(input: &str) -> String {
+  input
+    .lines()
+    .map(|line| {
+      if line.trim().starts_with('*') {
+        &line.trim()[1..]
+      } else {
+        line.trim()
+      }
+    })
+    .filter(|line| !line.is_empty())
+    .collect::<Vec<_>>()
+    .join("\n")
+}
+
+pub fn is_supported_doctest(path: &Path) -> bool {
+  let valid_ext = ["ts", "tsx", "js", "jsx"];
+  path
+    .extension()
+    .and_then(OsStr::to_str)
+    .map(|ext| valid_ext.contains(&ext) && !is_supported(path))
+    .unwrap_or(false)
+}
```

Comment:
Not really sure what's going on here, @iykekings please explain

Reply:
This is like opposite of is_supported for normal test files. It ensures doc tests are not run on test files and also that the file format is of normal expected type